### ISSUE TITLE
History completion is slow

### ISSUE THREAD

**Issue by LakeishaKowalczyk** [tridactyl/tridactyl#1016]
I have a 105MB places.sqlite LOL，This feature will make Firefox stuck（but the Built-in address bar will not）

`set historyresults 0` will Lead to an error: 
`# Error processing shim(history,search,[object Object]) Error: Type error for parameter query (Error processing maxResults: Integer 0 is too small (must be at least 1)) for history.search. `

**Comment by notJerl** [comment]
I've made a pull request for a minor change that'll make it just keep showing the top sites if you `set historyresults 0`.  Could probably be fixed better, but I'm not sure whether it'd be more desirable to do it this way or just not show autocompletions at all.

**Comment by LakeishaKowalczyk** [comment]
> 
> 
> I've made a pull request for a minor change that'll make it just keep showing the top sites if you `set historyresults 0`. Could probably be fixed better, but I'm not sure whether it'd be more desirable to do it this way or just not show autocompletions at all.

Thank you
Maybe it can be like this，generally not show autocompletions，when user type 'tab'，autocompletions will show

**Comment by bovine3dom** [comment]
That PR is merged (thanks @notJerl) but we are aware that our history performance is bad and would like to improve it.

**Comment by glacambre** [comment]
Related issues: https://github.com/tridactyl/tridactyl/issues/454, https://github.com/tridactyl/tridactyl/issues/229

**Comment by rob3c** [comment]
How can I completely disable history completions? Or even just completions more generally? Tridactyl can't keep up with my typing, and it just gets in the way.

**Comment by glacambre** [comment]
@robc You can disable `<Space>` not appearing where it should with `unbind --mode=ex <Space>`. As mentioned in @notJerl's comment, you can use `:set historyresults 0` in order to disable history completion. I don't think disabling other completions is possible.

**Comment by rob3c** [comment]
@glacambre Thanks for the unbind space trick!

Unfortunately, setting historyresults to 0 has no effect for me, which is why i jumped in eith my question. I can see results limited to a non zero number (still taking up the same vertical space though), but 0 just shows the default number of items.

**Comment by notJerl** [comment]
No, it has an effect.  If you set it to 0, it won't give you any _history_ completions.  It shows top sites, which is generated differently, by Firefox itself.  I think I could hack in a way to make it so if you set it to a negative number, it gives no suggestions whatsoever, but it would be very hacky and is getting into where it should honestly be put into the parser rewrite instead.  I also don't think it would actually be much faster.  Most of your slowdown was probably because how space is bound by default, which happens for all of ex-mode, not just history completion.

**Comment by rob3c** [comment]
@notJerl Thanks for the clarification! The little popup window still says `History` at the top, so I assumed it was history completions. However, as I look closer, I see it isn't really the history list per se, but must be the stupid Firefox top sites calculated history list that you mentioned (and which is apparently impossible to disable).

Now that spaces are in the right place, my main annoyance is that the window pops up at all for simple `:open` and `:tabopen`, because it often blocks part of the screen I'm looking at while typing. I have a lot of quicksearch bookmarks with shortcuts, and those don't seem to work with tabopen anyway. So maybe my best bet is remapping `t` to open and switch to a new tab immediately so I can leverage the Firefox address bar to use them.

Anyway, this is very minor stuff. Overall, I'm quite happy with such a nice alternative, now that my trusty vimperator of many years is dead. My muscle memory just needs some time to adjust :-)

**Comment by notJerl** [comment]
WebExtensions don't have access to bookmark shortcuts. For equivalent search shortcuts, you need to `set searchurls.whatever https://whatever.com/%s`, which will cause `tabopen whatever asdf` to open `https://whatever.com/asdf`.

**Comment by mozbugbox** [comment]
I'v done some profiling/timing on the history completion. It seems that the history/bookmark in `providers.ts` spends most of its time on the tridactyl proxy to the browser backend, which is `browserBg`.

If you use firefox `about:debugging` on tridactyl and timing the *bare metal* `browser.history.search()`, you'd found out it takes little time and it's much faster than through `browserBg.history.search()`.

```js
// time browser.history.search 
async function doit() {
  console.time("t1")
  await browser.history.search({text: "youtube", maxResults: 50, startTime: 0})
  console.timeEnd("t1")
}
await doit()
```

Similarly if you put `console.time/timeEnd()` around the `browserBg.history.search` in `providers.ts`, you'd see that the later was several times slower than the former. Maybe this behavior was expected but it explained why the history completion is slow.

Actually, if you type really fast, i.e. repeated history and bookmark search through `browserBg`, the proxy seemed get jamed somehow. The searching became really slow and hogged the CPU.

BTW, you can insert `console.time/timeEnd` directly into firefox debugger at the break point without changing the source. Just right click at the break point.

**Comment by bovine3dom** [comment]
Perhaps a connection based approach would make more sense? We could have a dedicated history search provider in the background that accepts filters one way and sends results the other way - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#Connection-based_messaging

**Comment by bovine3dom** [comment]
@LakeishaKowalczyk let's keep this open as I think the connection based approach is probably a good idea :)
